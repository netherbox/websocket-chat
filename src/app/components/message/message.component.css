:host {
  display: block;
  position: relative;
  padding: 12px 23px;
}

:host:hover {
  background: #f2f2f2;
}

:host:hover .tools {
  display: block;
}

.author {
  font-weight: 700;
}

.stamp {
  color: #9a9fa8;
  font-weight: normal;
}

.text {
  margin-top: 5px;
}

button,
input[type="submit"] {
  padding: 5px 10px;
  border: 1px solid #868d98;
  border-radius: 3px;
}

.tools {
  display: none;
  position: absolute;
  top: 12px;
  right: 23px;
}

.tools > * {
  margin-left: 5px;
}

.update .actions > * {
  margin-right: 5px;
}

.update input[type="text"] {
  display: block;
  padding: 12px 14px;
  border: 1px solid #868d98;
  margin: 16px 0 8px;
  border-radius: 3px;
}

.deleted {
  margin-top: 5px;
}
