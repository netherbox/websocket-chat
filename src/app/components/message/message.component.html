<div class="author">
  {{ message?.createdBy }}
  <span class="stamp">{{ message?.createdAt | date: "shortTime" }}</span>
</div>

<ng-container *ngIf="!message.deleted && !showUpdateForm">
  <div class="text">
    {{ message.text }}
  </div>

  <div class="tools" *ngIf="editable">
    <button (click)="onUpdateClick()">Edit</button>
    <button (click)="onDeleteClick()">Delete</button>
  </div>

  <div class="updated" *ngIf="message.updated">
    The message was edited by {{ message.updatedBy }} at
    {{ message.createdAt | date: "shortTime" }}
  </div>
</ng-container>

<ng-container *ngIf="showUpdateForm">
  <form class="update" (submit)="onUpdateMessageSubmit(updateMessageInput)">
    <input
      class="input"
      type="text"
      placeholder="Message"
      required
      #updateMessageInput
      [value]="message.text"
    />
    <div class="actions">
      <input type="submit" value="Submit" />
      <button (click)="onUpdateMessageCancel()">Cancel</button>
    </div>
  </form>
</ng-container>

<div class="deleted" *ngIf="message.deleted">
  The message was deleted by {{ message.deletedBy }} at
  {{ message.createdAt | date: "shortTime" }}
</div>
