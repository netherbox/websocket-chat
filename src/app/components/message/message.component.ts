import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { ChatMessage } from '../../models/chat-message.model';

@Component({
  selector: 'app-message',
  templateUrl: './message.component.html',
  styleUrls: ['./message.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ChatMessageComponent {
  @Input()
  message: ChatMessage = {
    messageId: '',
    text: '',
  };

  @Input()
  editable = false;

  showUpdateForm = false;

  @Output() update: EventEmitter<any> = new EventEmitter();

  @Output() delete: EventEmitter<any> = new EventEmitter();

  onUpdateClick() {
    this.showUpdateForm = true;
  }

  onUpdateMessageSubmit(messageInput: HTMLInputElement): boolean {
    this.update.emit({ ...this.message, text: messageInput.value });
    this.showUpdateForm = false;
    return false;
  }

  onUpdateMessageCancel() {
    this.showUpdateForm = false;
  }

  onDeleteClick() {
    this.delete.emit({ ...this.message });
  }
}
