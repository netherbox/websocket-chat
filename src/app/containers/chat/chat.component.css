:host {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
}

.messages {
  flex-grow: 1;
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-direction: column;
  margin-top: auto;
  overflow: auto;
  height: 1px;
}

.messages > li {
  padding: 10px 14px;
}

.messages > li:first-child {
  margin-top: auto !important;
}

.form {
  display: flex;
}

.input {
  flex-grow: 1;
  display: block;
  padding: 12px 14px;
  border: 1px solid #868d98;
  margin: 16px;
  border-radius: 3px;
}
