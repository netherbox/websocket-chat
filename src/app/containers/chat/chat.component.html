<ul class="messages" #messages>
  <li *ngFor="let message of chatService.messages$ | async">
    <div class="author">
      {{ message.createdBy }}
      <span class="stamp">{{ message.createdAt | date: "shortTime" }}</span>
    </div>
    <div class="text">
      {{ message.text }}
    </div>
    <div class="tools" *ngIf="message.createdBy == chatService.currentUserId">
      <button>Edit</button> &#160; <button>Delete</button>
    </div>
  </li>
</ul>

<form class="form" (submit)="onSendMessageSubmit(messageInput)">
  <input class="input" placeholder="Message" required #messageInput />
</form>
