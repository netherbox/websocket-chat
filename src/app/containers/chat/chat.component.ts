import {
  AfterViewInit,
  Component,
  ElementRef,
  OnDestroy,
  ViewChild,
} from '@angular/core';
import { Subscription } from 'rxjs';

import { ChatMessage } from '../../models/chat-message.model';
import { ChatService } from '../../services/chat.service';

@Component({
  selector: 'app-chat',
  templateUrl: './chat.component.html',
  styleUrls: ['./chat.component.css'],
})
export class Chat implements AfterViewInit, OnDestroy {
  @ViewChild('messages') messages: ElementRef | undefined;

  updatingMessage: ChatMessage | undefined;

  messagesSubcription: Subscription;

  constructor(public chatService: ChatService) {
    this.messagesSubcription = this.chatService.messages$.subscribe(() =>
      this.scrollDown()
    );
  }

  ngAfterViewInit(): void {
    this.scrollDown();
  }

  onSendMessageSubmit(messageInput: HTMLInputElement): boolean {
    this.sendMessage(messageInput.value);
    messageInput.value = '';
    return false;
  }

  async sendMessage(text: string) {
    await this.chatService.send(text);
  }

  scrollDown(): void {
    setTimeout(() => {
      if (this.messages) {
        this.messages.nativeElement.scrollTop =
          this.messages.nativeElement.scrollHeight;
      }
    }, 100);
  }

  onUpdateClick(message: ChatMessage) {
    this.updatingMessage = { ...message };
  }

  onUpdateMessageSubmit(messageInput: HTMLInputElement): boolean {
    if (this.updatingMessage) {
      this.updateMessage({ ...this.updatingMessage, text: messageInput.value });
      this.updatingMessage = undefined;
    }
    return false;
  }

  async updateMessage(message: ChatMessage) {
    await this.chatService.update(message);
  }

  onUpdateMessageCancel() {
    this.updatingMessage = undefined;
  }

  onDeleteClick(message: ChatMessage) {
    this.chatService.delete(message);
  }

  messageById(_index: number, message: ChatMessage) {
    return message.messageId;
  }

  ngOnDestroy(): void {
    this.messagesSubcription?.unsubscribe();
  }
}
