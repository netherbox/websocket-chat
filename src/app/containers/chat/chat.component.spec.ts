import { ComponentFixture, TestBed } from '@angular/core/testing';
import { BehaviorSubject } from 'rxjs';

import { ChatComponent } from './chat.component';
import { ChatService } from '../../services/chat.service';
import { ChatMessageComponent } from '../../components/message/message.component';
import { ChatMessage } from 'src/app/models/chat-message.model';

describe('ChatComponent', () => {
  let testMessage: ChatMessage;
  let component: ChatComponent;
  let fixture: ComponentFixture<ChatComponent>;
  let messagesMock$: BehaviorSubject<boolean>;
  let chatService: ChatService;

  beforeEach(async () => {
    testMessage = {
      messageId: 'test-messageId',
      text: 'test-text',
    };

    messagesMock$ = new BehaviorSubject<boolean>(false);

    await TestBed.configureTestingModule({
      declarations: [ChatComponent, ChatMessageComponent],
      providers: [
        {
          provide: ChatService,
          useValue: {
            messages$: messagesMock$,
            send: jest.fn(),
            update: jest.fn(),
            delete: jest.fn(),
          },
        },
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(ChatComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    chatService = TestBed.inject(ChatService);
  });

  it('should successfully create component with empty list', () => {
    expect(component).toBeTruthy();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should set messagesSubcription on creation', () => {
    expect(component.messagesSubcription).toBeTruthy();
  });

  it('should call scroll down after view initialized', () => {
    jest.spyOn(component, 'scrollDown').mockImplementation();

    component.ngAfterViewInit();

    expect(component.scrollDown).toHaveBeenCalled();
  });

  describe('scrollDown', () => {
    it('should call scroll down after view initialized', () => {
      jest.useFakeTimers();

      component.messagesContainer = {
        nativeElement: {
          scrollTop: 0,
          scrollHeight: 1000,
        },
      };

      component.scrollDown();

      jest.runAllTimers();

      expect(component.messagesContainer.nativeElement.scrollTop).toEqual(1000);
    });

    it('should do nothing if messagesContainer is not initialized', () => {
      jest.useFakeTimers();

      component.messagesContainer = undefined;

      component.scrollDown();

      jest.runAllTimers();

      expect(component.messagesContainer).toBeUndefined();
    });
  });

  describe('onSendMessageSubmit', () => {
    it('should   view initialized', () => {
      jest.spyOn(component, 'sendMessage');

      const mockInput = {
        value: 'test-input-value',
      } as unknown as HTMLInputElement;

      component.onSendMessageSubmit(mockInput);

      expect(component.sendMessage).toHaveBeenCalledWith('test-input-value');
      expect(mockInput.value).toEqual('');
    });
  });

  describe('messageById', () => {
    it('should return message id to be used by ngFor trackBy', () => {
      const result = component.messageById(0, testMessage);

      expect(result).toEqual(testMessage.messageId);
    });
  });

  describe('sendMessage', () => {
    it('should call send method of chatService', async () => {
      jest.spyOn(chatService, 'send').mockResolvedValue();

      await component.sendMessage('test-text');

      expect(chatService.send).toHaveBeenCalledWith('test-text');
    });
  });

  describe('onUpdateMessage', () => {
    it('should call update method of chatService', async () => {
      jest.spyOn(chatService, 'update').mockResolvedValue();

      await component.onUpdateMessage(testMessage);

      expect(chatService.update).toHaveBeenCalledWith(testMessage);
    });
  });

  describe('onDeleteMessage', () => {
    it('should call delete method of chatService', async () => {
      jest.spyOn(chatService, 'delete');

      component.onDeleteMessage(testMessage);

      expect(chatService.delete).toHaveBeenCalledWith(testMessage);
    });
  });

  describe('ngOnDestroy', () => {
    it('should unsubscribe from messagesSubcription', async () => {
      jest.spyOn(component.messagesSubcription, 'unsubscribe');

      component.ngOnDestroy();

      expect(component.messagesSubcription.unsubscribe).toHaveBeenCalled();
    });
  });
});
