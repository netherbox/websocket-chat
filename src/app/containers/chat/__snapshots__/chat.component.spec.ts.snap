// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatComponent should render messages 1`] = `
<div
  id="root1"
>
  <ul
    class="messages"
  >
    
    <li>
      <app-message>
        <div
          class="author"
        >
           test-currentUserId 
          <span
            class="stamp"
          >
            
          </span>
        </div>
        
        
        <div
          class="text"
        >
           test-text-1 
        </div>
        
        <div
          class="tools"
        >
          <button>
            Edit
          </button>
          <button>
            Delete
          </button>
        </div>
        
        
        
        
      </app-message>
    </li>
    <li>
      <app-message>
        <div
          class="author"
        >
           test-someOtherUserId 
          <span
            class="stamp"
          >
            
          </span>
        </div>
        
        
        <div
          class="text"
        >
           test-text-2 
        </div>
        
        
        
        
        
      </app-message>
    </li>
  </ul>
  <form
    class="form"
  >
    <input
      class="input"
      placeholder="Message"
      required=""
      type="text"
    />
  </form>
</div>
`;

exports[`ChatComponent should successfully create component with empty list 1`] = `
<div
  id="root0"
>
  <ul
    class="messages"
  >
    
  </ul>
  <form
    class="form"
  >
    <input
      class="input"
      placeholder="Message"
      required=""
      type="text"
    />
  </form>
</div>
`;
